### PR TITLE
feat(SmartAI): import SMART_EVENT_FLAG_WHILE_CHARMED from TC

### PR DESCRIPTION
**Pull request by FrancescoBorzi**
**NOTE: this PR needs to be carefully tested since it affects many things**

---------

This PR implements `SMART_EVENT_FLAG_WHILE_CHARMED` by importing/adapting the following commits from TC (by Treeston and Ariel-):

* [TrinityCore/TrinityCore@96f9038](https://github.com/TrinityCore/TrinityCore/commit/96f90381e3ed696861354ea4ddf4ac0c495031c5)
* [TrinityCore/TrinityCore@a4623ef](https://github.com/TrinityCore/TrinityCore/commit/a4623efaa67be1cfea19e561be83f118a784b9d1)
* [TrinityCore/TrinityCore@3e59637](https://github.com/TrinityCore/TrinityCore/commit/3e596376a7fae9fd585fac12e5122fd8f1b7ce14)
* [TrinityCore/TrinityCore@0f7efc8](https://github.com/TrinityCore/TrinityCore/commit/0f7efc86d066ed614fa7b192da81c48d586f8b85)

- Closes: https://github.com/azerothcore/azerothcore-wotlk/issues/731
- Updates https://github.com/azerothcore/azerothcore-wotlk/issues/1372

## Original description:

Original PR here: https://github.com/TrinityCore/TrinityCore/pull/17738

Scripts/SmartScripts: Fix charmed behavior for SmartAI. 

- Core AI logic will no longer issue attack/movement commands while under player control.
- Add new `SMART_EVENT_FLAG_WHILE_CHARMED` (`0x200`). Any event without this flag won't run while charmed.
- Add `SMART_EVENT_FLAG_WHILE_CHARMED` to existing actions, except a select subset (movement, talking, and nontriggered casts, pretty much)

Note for quest fixes: **no claim to completeness**

## How to test

Basically this PR changes the default behaviour of all smart scripts except for the ones marked as   `SMART_EVENT_FLAG_WHILE_CHARMED` . So it needs to be properly checked.

To have an idea about what needs to be tested, I recommend to:

- Read the original PR discussions (as well as comments in the imported commits)
- Carefully check the SQL file introduced in this PR to better understand what's going on

### CODE REVIEW

**Comment by sucofog** [comment]
How this can be tested @FrancescoBorzi ?

**Comment by FrancescoBorzi** [comment]
@sucofog The most important thing is to avoid any regression - one way could be trying different existing SmartAI and make sure that they are not broken with this commit.

Any feedback is appreciated.

**Comment by sucofog** [comment]
Tested some involved SAI quests involving the addition of the SMART_EVENT_FLAG_WHILE_CHARMED flags, it's fine to merge changes for me

**Comment by comix1988** [comment]
I tried to test this PR even if I didn't understand what to test precisely. There are my test:

- Hunter & lokko pet ( aggressive, defence and passive stance)
- taxi node ( from dalaran to crystalsong & from K3 to dalaran)
- some quest:
                      - Battling the elements;
                      - spy hunter;
                      - Generosity abounds;
                      - Seed of chaos;
                      - The kessel run;
                      - The drakkensryd;
                      - The black knight's orders;
                      - Reconnaissance flight;
                    **- Escape from silverbrook;
                      - Krolmir, hammer of stroms;**
The last two quests work on the master branch but do not last the PR test:
-Escape from silverbrook: sometimes the horse doesn't start its run, i had to get on and off to get it started
- Krolmir, hammer of storms: when you talk to the npc "King Jokkum" the npc that transports you is spawned but doesn't work

**Comment by poszer** [comment]
More than 20 days without any issue with this commit applied, for now. All quest looks fine for me, but let's wait more until this merged. It's best to test this on the creatures which's are called from the item for example. So everyone who is free, and able to test. Try to research for these quest type's :dango:

**Comment by FrancescoBorzi** [comment]
@poszer thanks for the feedback. What about the 2 quests pointed by @comix1988  (in bold) that appear to be broken by this PR ?

Maybe @wetbrownsauce can also help here, since this is a big change

**Comment by gengarshadowball** [comment]
why was this closed? Is it implemented now?

**Comment by FrancescoBorzi** [comment]
@gengarshadowball because it was breaking some stuff

**Comment by gengarshadowball** [comment]
Do you need help figuring it out?

**Comment by FrancescoBorzi** [comment]
@gengarshadowball that would be nice, yes. However we can only merge this if we are 100% sure that it's not breaking stuff

**Comment by zqgmzq** [comment]
maybe some body can continue this work?
it is necessary pr for actual quests

**Comment by FrancescoBorzi** [comment]
@zqgmzq we need more help with testing PRs so we can afford to import more stuff like this.

Please help us https://www.azerothcore.org/wiki/how-to-test-a-pr

**Comment by zqgmzq** [comment]
my team can help with testing
can u reopen and update this pr?

**Comment by FrancescoBorzi** [comment]
> my team can help with testing can u reopen and update this pr?

hi @zqgmzq there are currently more than 30 PRs that need to be tested, if your team can help that would be very appreciated. You can always find the list of PRs that need to be tested here:

https://github.com/azerothcore/azerothcore-wotlk/pulls?q=is%3Apr+is%3Aopen+label%3A%22Waiting+to+be+Tested%22

so we can't afford spending time on other features before the testing workload is reduced, because we don't have enough time so please help us to speed up the development pace
